company_from_url/models/res_users.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 AppsToGROW - Henrik Norlin
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, exceptions, fields, models, SUPERUSER_ID, tools
+from odoo.http import request
+from odoo.tools import mute_logger
+
+from psycopg2 import ProgrammingError
+
+import logging
+_logger = logging.getLogger(__name__)
+
+import odoo.sql_db
+
+
+class ResUsers(models.Model):
+    _inherit = 'res.users'
+
+    def _inverse_company_id(self):
+        """ Save selected company id to stored_company_id """
+        for this in self:
+            this.stored_company_id = this.company_id
+
+    def _compute_company_id(self):
+        """ Change company_id on user to a computed field based on URL """
+        company_obj = self.env['res.company']
+        company_id = self.env.context.get('force_company', None)
+        company = company_id and company_obj.browse(company_id)
+        if not company:
+            try:
+                url_domain = request.httprequest.host.partition(":")[0]
+                company = company_obj.search([
+                    ('access_url', '=', url_domain)
+                ])[:1] & self.env.user.company_ids
+            except RuntimeError:
+                company = None
+        for this in self:
+            this.company_id = \
+                company or this.stored_company_id or this.company_ids[:1]
+
+    company_id = fields.Many2one(
+        compute='_compute_company_id',
+        inverse='_inverse_company_id',
+        string='Company (from URL)',
+        store=False, index=False, required=False)
```

Comment:
this will cause you a lot of trouble because there will be 10k scripts/addons that expect the database column to exist. Better override `read` and if `request` is not falsy (that's much cheaper to check than caching the runtime error), set your company in the result dict. That's cached within the current environment which should be good enough for starters, you could also set this in some property in the request, then for every request you only search once.

Reply:
I have [tried](https://github.com/sunflowerit/multi-company/pull/1/files) to do this, but for some reason the cached values of user.company_id bite me in the case of using a `read` override, but not when using a computed field.

Reply:
I'll debug more, the caching issue was located [here](https://github.com/OCA/OCB/blob/10.0/addons/web/models/ir_http.py#L38)

The `stored_company_id` field is unnecessary, it was intended more like a feature, so you can still see which company the user would get if he arrives through a non-company-forcing domain name.

Reply:
Finally got this; `read_from_database` was already filling the cache, so I had to override that instead of `read`.

Reply:
huh, sorry, didn't think of this. Thanks for the pointer